### PR TITLE
#12985: Expose `ttnn::ccl::Topology` at python level

### LINKED ISSUE #12985: Expose ttnn::ccl::Topology at python level
As a step for unifying CCL calls and cleaning up the interface for calling from TTNN, the enum must be exposed to the TTNN level.

This issue is to track that work. For the time being, we should only expose `Line` and `Ring` topologies. The enum should be scoped.

### PR DESCRIPTION
**Pull request by Aswinmcw**
Tracking Issue : #12985 

Expose ccl Ring and Linear topology to be accessed from python

### Checklist
- [x] Post commit CI passes - https://github.com/tenstorrent/tt-metal/actions/runs/10991198137
- [ ] Blackhole Post commit (if applicable)
- [ ] Model regression CI testing passes (if applicable)
- [ ] Device performance regression CI testing passes (if applicable)
- [ ] New/Existing tests provide coverage for changes

### CODE REVIEW

**Comment by cfjchu** [comment]
If we're exposing this enum and adding this ccl::Topology as a new function parametrization, does it make sense to consolidate `ttnn.all_gather` and `ttnn.line_all_gather`?

**Comment by SeanNijjar** [comment]
> If we're exposing this enum and adding this ccl::Topology as a new function parametrization, does it make sense to consolidate `ttnn.all_gather` and `ttnn.line_all_gather`?

Yes - this is one of the immediate follow on items as a result

**Comment by SeanNijjar** [comment]
Hi @Aswinmcw, apologies as I forgot there was still a separate topology enum (legacy) for All-gather. 


Can you please instead use and pass `enum Topology { Ring = 0, Linear = 1, Meash = 2 };` Meash has a typofrom `ttnn/cpp/ttnn/operations/ccl/ccl_host_datastructures.hpp`? All-gather should also be using this topology enum and the old one should just be deleted.

File: ttnn/cpp/ttnn/operations/ccl/all_gather/all_gather.hpp [hunk 1]
```diff
@@ -5,19 +5,25 @@
...
 
 #include "ttnn/decorators.hpp"
+#include "ttnn/operations/ccl/ccl_host_datastructures.hpp"
 
 namespace ttnn {
 namespace operations {
 namespace ccl {
 
+namespace all_gather_op {
+using ttnn::ccl::Topology;
```
Comment (SeanNijjar) [inline ttnn/cpp/ttnn/operations/ccl/all_gather/all_gather.hpp:15]: I'm confused why this was namespaced to `all_gather_op`?

**Review by SeanNijjar — CHANGES_REQUESTED** [review]
High level this is looking okay but we shouldn't be sending the Topology to the all_gather_op namespace as this topology is/will be shared across all CCL ops.

File: ttnn/cpp/ttnn/operations/ccl/all_gather/all_gather.hpp (hunk 1 above)
Comment (Aswinmcw) [reply ttnn/cpp/ttnn/operations/ccl/all_gather/all_gather.hpp:15]: I have done this as the same method is followed in `all_gather_op.hpp`, I will revert this change to call directly without adding additional namespace

File: ttnn/cpp/ttnn/operations/ccl/all_gather/all_gather.cpp [hunk 2]
```diff
@@ -8,8 +8,8 @@
 
 namespace ttnn::operations::ccl {
 
-ttnn::Tensor ExecuteAllGather::invoke(const ttnn::Tensor& input_tensor, const uint32_t dim, const uint32_t num_links, const std::optional<ttnn::MemoryConfig>& memory_config, const std::optional<size_t> num_workers, const std::optional<size_t> num_buffers_per_channel) {
-    return ttnn::operations::ccl::all_gather(input_tensor, dim, num_links, memory_config, num_workers, num_buffers_per_channel);
+ttnn::Tensor ExecuteAllGather::invoke(const ttnn::Tensor& input_tensor, const uint32_t dim, const uint32_t num_links, const std::optional<ttnn::MemoryConfig>& memory_config, const std::optional<size_t> num_workers, const std::optional<size_t> num_buffers_per_channel, const all_gather_op::Topology topology) {
```
Comment (SeanNijjar) [inline ttnn/cpp/ttnn/operations/ccl/all_gather/all_gather.cpp:11]: This is still in the all gather namespace? It should just be ttnn::ccl::Topology

(same comment applies elsewhere in the commit still... Am I seeing old changes?)

**Review by SeanNijjar — CHANGES_REQUESTED** [review]

File: ttnn/cpp/ttnn/operations/ccl/all_gather/all_gather_pybind.cpp [hunk 3]
```diff
@@ -60,6 +67,7 @@ void py_bind_all_gather(pybind11::module& module) {
             * :attr:`memory_config` (Optional[ttnn.MemoryConfig]): Memory configuration for the operation.
             * :attr:`num_workers` (int): Number of workers to use for the operation.
             * :attr:`num_buffers_per_channel` (int): Number of buffers per channel to use for the operation.
+            * :attr:`topology`: Topology to be used for the operation.
```
Comment (SeanNijjar) [inline ttnn/cpp/ttnn/operations/ccl/all_gather/all_gather_pybind.cpp:70]: Can you please add something like " Allowable options are Line and Ring

**Review by SeanNijjar — APPROVED** [review]
Looks good (one more minor docs change requested)! Please also run t3000 frequent to ensure no CCL tests were unintentionally affected.

**Review by eyonland — APPROVED** [review]
I recommend that we also allow the namespace to include ttnn::Topology so that it aligns with the python ttnn.Topology.   At the moment it appears that the code only supports ttnn::ccl::Topology.

**Pull request merged**